OpenSim/Common/Test/testTableSource.cpp

```diff
@@ -0,0 +1,216 @@
+
+class TheWorld : public OpenSim::Component {
+    OpenSim_DECLARE_CONCRETE_OBJECT(TheWorld, Component);
+public:
+    TheWorld() : Component() {
+        // Constructing own properties, connectors, inputs or connectors? Must 
+        // invoke!
+        constructInfrastructure();
+    }
+
+    TheWorld(const std::string& fileName, bool updFromXMLNode = false)
+        : Component(fileName, updFromXMLNode) {
+        // have to construct this Component's properties so that deserialization
+        // from XML has a place to go.
+        constructInfrastructure();
+        // Propagate XML file values to properties 
+        updateFromXMLDocument();
+        // add components listed as properties as sub components.
+        finalizeFromProperties();
+    }
+
+    void add(Component* comp) {
+        addComponent(comp);
+        // Edit Sub 
+        /*Sub& subc = */updMemberSubcomponent<Sub>(intSubix);
+    }
+
+    // Top level connection method for this all encompassing component, TheWorld
+    void connect() {
+        Super::connect(*this);
+    }
+    void buildUpSystem(SimTK::MultibodySystem& system) { 
+        connect();
+        addToSystem(system);
+    }
+
+    const SimTK::SimbodyMatterSubsystem& getMatterSubsystem() const { 
+        return *matter; 
+    }
+
+    SimTK::SimbodyMatterSubsystem& updMatterSubsystem() const { 
+        return *matter; 
+    }
+
+    const SimTK::GeneralForceSubsystem& getForceSubsystem() const { 
+        return *forces; 
+    }
+
+    SimTK::GeneralForceSubsystem& updForceSubsystem() const { 
+        return *forces; 
+    }
+
+protected:
+    // Component interface implementation
+    void extendAddToSystem(SimTK::MultibodySystem& system) const override {
+        if (system.hasMatterSubsystem()){
+            matter = system.updMatterSubsystem();
+        }
+        else{
+            // const Sub& subc = getMemberSubcomponent<Sub>(intSubix);
+
+            SimTK::SimbodyMatterSubsystem* old_matter = matter.release();
+            delete old_matter;
+            matter = new SimTK::SimbodyMatterSubsystem(system);
+
+            SimTK::GeneralForceSubsystem* old_forces = forces.release();
+            delete old_forces;
+            forces = new SimTK::GeneralForceSubsystem(system);
+
+            SimTK::Force::UniformGravity gravity(*forces, *matter, 
+                                                 SimTK::Vec3(0, -9.816, 0));
+            fix = gravity.getForceIndex();
+
+            system.updMatterSubsystem().setShowDefaultGeometry(true);
+        }
+    }
+
+private:
+    // Keep track of pointers to the underlying computational subsystems. 
+    mutable SimTK::ReferencePtr<SimTK::SimbodyMatterSubsystem> matter;
+    mutable SimTK::ReferencePtr<SimTK::GeneralForceSubsystem> forces;
+
+    // keep track of the force added by the component
+    mutable SimTK::ForceIndex fix;
+
+    MemberSubcomponentIndex intSubix{constructSubcomponent<Sub>("internalSub")};
+}; // class TheWorld
+
+
```

Comment:
Although, I like having a separate test, I am not a fan of the duplication of `TheWorld` and its dependencies here. Would it not be more efficient to add additional cases (SUBTEST) for testing the `TableSource` to `testComonentInterface.cpp` which already provides the necessary "environment". My fear is that we have to maintain the same environment in two places.